File: src/cold/serial_handler_node.py
#!/usr/bin/env python3

import rospy
from pymodbus.client import ModbusSerialClient as ModbusClient
from motor_control_pkg.srv import ModbusWrite, ModbusWriteResponse  # カスタムサービスのインポート
import serial
import threading

class SerialHandlerNode:
    def __init__(self):
        # Modbus Client の初期化
        self.client = ModbusClient(
            method='rtu',
            port='/dev/ttyUSB0',
            baudrate=115200,
            timeout=3,
            parity=serial.PARITY_EVEN,
            stopbits=serial.STOPBITS_ONE
        )
        self.lock = threading.Lock()  # ロックを使ってアクセスを制御

        if not self.client.connect():
            rospy.logerr("Failed to connect to Modbus server")
            exit(1)

        # サービスの定義
        self.service = rospy.Service('modbus_write', ModbusWrite, self.handle_write_request)

    def handle_write_request(self, req):
        with self.lock:
            try:
                # リクエストの内容を取得
                address = req.address
                data = req.data
                slave_id = req.slave_id

                result = self.client.write_registers(address=address, values=data, unit=slave_id)

                # レスポンスの生成
                if result.isError():
                    return ModbusWriteResponse(success=False, message=str(result))
                else:
                    return ModbusWriteResponse(success=True, message="Write successful")
            except Exception as e:
                rospy.logerr(f"Modbus write failed: {e}")
                return ModbusWriteResponse(success=False, message=str(e))

    def run(self):
        rospy.spin()

if __name__ == "__main__":
    rospy.init_node('serial_handler_node')
    node = SerialHandlerNode()
    node.run()

File: src/motor_control_pkg/src/serial_handler_node.cpp
#include <ros/ros.h>
#include <motor_control_pkg/ModbusWrite.h>
#include <modbus/modbus.h>
#include <mutex>
#include <vector>
#include <cstdint>

class SerialHandlerNode {
public:
    SerialHandlerNode() : last_modbus_time(ros::Time::now()) {
        // Modbus client の初期化
        ctx = modbus_new_rtu("/dev/ttyUSB0", 115200, 'E', 8, 1); // パリティは Even, ストップビットは 1
        modbus_set_response_timeout(ctx, 1, 0); // タイムアウト1秒

        if (ctx == NULL) {
            ROS_ERROR("Unable to create the Modbus context");
            exit(1);
        }

        if (modbus_connect(ctx) == -1) {
            ROS_ERROR("Failed to connect to the Modbus server");
            modbus_free(ctx);
            exit(1);
        }

        // サブスクライバーを作成
        sub = nh.subscribe("modbus_request", 1000, &SerialHandlerNode::handleModbusRequest, this);
    }

    void handleModbusRequest(const motor_control_pkg::ModbusWrite::ConstPtr& req) {
        std::unique_lock<std::mutex> lock(modbus_mutex);

        // Modbus slave IDの設定
        int rc = modbus_set_slave(ctx, req->slave_id);
        if (rc == -1) {
            ROS_ERROR("Failed to set Modbus slave ID");
            return;
        }

        // リクエストされたデータを書き込む
        std::vector<uint16_t> data(req->data.size());
        for (size_t i = 0; i < req->data.size(); ++i) {
            data[i] = static_cast<uint16_t>(req->data[i]);  // int から uint16_t へキャスト
        }

        // Modbusレジスタに書き込み
        rc = modbus_write_registers(ctx, req->address, data.size(), data.data());
        if (rc == -1) {
            ROS_ERROR("Modbus write failed: %s", modbus_strerror(errno));
            return;
        }

        ROS_INFO("Modbus write successful");
    }

    void run() {
        ros::spin();
    }

    ~SerialHandlerNode() {
        modbus_close(ctx);
        modbus_free(ctx);
    }

private:
    ros::NodeHandle nh;
    ros::Subscriber sub;
    modbus_t *ctx; // Modbusコンテキスト
    std::mutex modbus_mutex; // スレッドセーフのためのミューテックス
    ros::Time last_modbus_time; // 最後にModbus通信を行った時間
};

int main(int argc, char **argv) {
    ros::init(argc, argv, "serial_handler_node");
    SerialHandlerNode node;
    node.run();
    return 0;
}
